File: 0503-next-greater-element-ii/0503-next-greater-element-ii.java
class Solution {
    public int[] nextGreaterElements(int[] nums) {
        int [] res = new int[nums.length] ;
        Stack<Integer> stack = new Stack<>() ;
        int l = nums.length ;
        for (int i = l -1 ; i >=0 ; i--){
            while(! stack.isEmpty() && nums[i] >=stack.peek()) stack.pop() ;
            stack.push(nums[i]) ;
        }
        for (int i = l -1 ; i >= 0 ; i--){
            int ele = -1 ;
            while(! stack.isEmpty() && nums[i] >= stack.peek()) stack.pop() ;
            if (! stack.isEmpty()) ele = stack.peek() ;
            stack.push(nums[i]) ;
            res[i] = ele ;
        }
        return res ;
        
        
    }
}

File: 0503-next-greater-element-ii/0503-next-greater-element-ii.py
class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        l = len(nums) 
        result = [-1] * l 
        stack = []
        for i in range( l -1 , -1 , -1) :
            while stack and nums[i] >= stack[-1] :
                stack.pop()
            stack.append(nums[i])
        for i in range(l - 1 , -1 ,-1 ) :
            while(stack and nums[i] >= stack[-1]) :
                stack.pop()
            if stack :
                result[i] = stack[-1]
            stack.append(nums[i]) 
        return result 

        